docs/design/gang_scheduling.md

```diff
@@ -0,0 +1,631 @@
+This process continues until there are no more placeholder pods to be allocated.
+
+The shim at that point should create the AllocationAsk for the real pod(s) that it has buffered.
+The core cannot and must not assume that there is only one task group per application.
+The core is also not in the position to assume that it has received all AllocationAsks that belong to the task group if option 1 as described above is used by a shim.
+This is also why we have the assumption that every pod creates a placeholder request to the core.
+
+The second change is the replacement of the placeholder pods with the real pods.
+The shim creates an AllocationAsk with the _taskGroupName_ set but the _placeholder_ flag is not set.
+
+The process described here lines up with the process for generic pre-emption.
+An allocation is released by the core and then confirmed by the shim.
+For gang scheduling we have a simple one new to one release relation in the case of pre-emption we can use the same flow with a one new to multiple release relation.
+
+The scheduler processes the AllocationAsk as follows:
+1. Check if the application has an unreleased allocation for a placeholder allocation with the same _taskGroupName._ If no placeholder allocations are found a normal allocation cycle will be used to allocate the request.
+2. A placeholder allocation is selected and marked for release. A request to release the placeholder allocation is communicated to the shim. This must be an async process as the shim release process is dependent on the underlying K8s response which might not be instantaneous.  
+   NOTE: no allocations are released in the core at this point in time.
+3. The core “parks” the processing of the real AllocationAsk until the shim has responded with a confirmation that the placeholder allocation has been released.  
+   NOTE: locks are released to allow scheduling to continue
+4. After the confirmation of the release is received from the shim the “parked” AllocationAsk processing is finalised.
+5. The AllocationAsk is allocated on the same node as the placeholder used.
+   The removal of the placeholder allocation is finalised in either case. This all needs to happen as one update to the application, queue and node.
+    * On success: a new Allocation is created.
+    * On Failure: try to allocate on a different node, if that fails the AllocationAsk becomes unschedulable triggering scale up. 
+6. Communicate the allocation back to the shim (if applicable, based on step 5)
+
+
+## Application completion
+Application completion has been a long standing issue.
+Currently, applications do not transition to a _completed_ state when done.
+The current states for the application are [documented here](./scheduler_object_states.md).
+However, at this point in time an application will not reach the _completed_ state and will be stuck in _waiting_.
+
+This provides a number of issues specifically around memory usage and cleanup of queues in long running deployments.
```

Comment:
I think this needs to be updated. Now we have introduced `Completed`, `Completing`, `Failed`, and `Failing` state. So the app completion is no longer an issue. 
The second thing is: what does this sentence mean here?

```
However, at this point in time an application will not reach the _completed_ state and will be stuck in _waiting_.
```

Reply:
That was done after design was written. We should not change the design doc to fit the implementation afterwards.

Reply:
Not the exact point, but still part of the design. Having the design doc saying something but the implementation does something else seems to be a problem to me.

Reply:
The failing & completing state were introduced to fix issues in the cleanup. That was done on top of what we had designed. We also do not go back and now fix the gang scheduling design after fixing YUNIKORN-639 which also changes the flow of the states.
A design you write, implement and then on top of the design you fix issues. You do not go back then and re-write the design doc. You update the user docs. There is a clear distinction between the two.

Reply:
IMO, we need to keep this updated if the design changes.
Otherwise, what's the point to have an out-of-dated doc here, people will get confused when they look into this doc and trying to understand what has been implemented.